boot/bootstate20.go

```diff
@@ -244,6 +253,50 @@ func (ks20 *bootState20Kernel) commit() error {
 	return nil
 }
 
+// chooseAndCommitSnapInitramfsMount chooses which snap should be mounted
+// during the early boot sequence, i.e. the initramfs, and commits that
+// choice if it needs state updated.
+// Choosing to boot/mount the base snap needs to be committed to the
+// modeenv, but no state needs to be committed when choosing to mount a
+// kernel snap.
+func (ks20 *bootState20Kernel) chooseAndCommitSnapInitramfsMount() (sn snap.PlaceInfo, err error) {
+	err = ks20.kModeenv.loadModeenv()
+	if err != nil {
+		return nil, err
+	}
+
+	// first do the generic choice of which snap to use
+	first, second, err := genericEarlyBootChooseSnap(ks20, TryingStatus, "kernel")
+	if err != nil {
+		return nil, err
+	}
+
+	// now validate the chosen kernel snap against the modeenv CurrentKernel's
+	// setting
+	// make a map to easily check if a kernel snap is valid or not
+	validKernels := make(map[string]bool, len(ks20.kModeenv.modeenv.CurrentKernels))
+	for _, validKernel := range ks20.kModeenv.modeenv.CurrentKernels {
+		validKernels[validKernel] = true
+	}
+
+	if !validKernels[first.Filename()] {
+		if second != nil {
+			// try the second snap, which will always be the fallback snap if
+			// set
+			if !validKernels[second.Filename()] {
+				return nil, fmt.Errorf("fallback kernel snap %q is not trusted in the modeenv", second.Filename())
+			}
+			// log message that first (try snap) wasn't trusted
+			logger.Noticef("cannot use try-kernel snap: %q is not trusted in the modeenv", first.Filename())
+			return second, nil
```

Comment:
will snapd in user space do the right thing in this case?

Reply:
hmm? I don't understand the question as snapd in user space should never call this function since it's specifically only for the initramfs, hence it's just for `snap-bootstrap initramfs-mounts` to use

Reply:
I mean we are picking a different outcome without manipulating the status,  will snapd notice this?

Reply:
TLDR, no snapd doesn't notice this. What we need to do actually is reboot instead of falling back here. We could manipulate the status and let snapd in user-space figure it out, but we won't necessarily get that far, because we could have incompatible kernel modules between the fallback kernel we mounted and the try kernel we booted. The other thing about doing that is that snapd would also need to detect that the mounted snap and kernel snap are out of sync and do a reboot to fix the situation, which is not ideal.